synapse/handlers/register.py

```diff
@@ -717,21 +738,40 @@ class and RegisterDeviceReplicationServlet.
                 )
             valid_until_ms = self.clock.time_msec() + self.session_lifetime
 
+        refresh_token = None
+        refresh_token_id = None
+
         registered_device_id = await self.device_handler.check_device_registered(
             user_id, device_id, initial_display_name
         )
         if is_guest:
             assert valid_until_ms is None
             access_token = self.macaroon_gen.generate_guest_access_token(user_id)
         else:
+            if should_issue_refresh_token:
+                (
+                    refresh_token,
+                    refresh_token_id,
+                ) = await self._auth_handler.get_refresh_token_for_user_id(
+                    user_id,
+                    device_id=registered_device_id,
+                )
+                valid_until_ms = self.clock.time_msec() + self.access_token_lifetime
```

Comment:
it seems like we need to define how `session_lifetime` and `access_token_lifetime`  interact. Currently it looks like `session_lifetime` is just ignored if the client requests a refresh token, which isn't really any good.

`session_lifetime`'s main purpose is to force a periodic re-authentication with an external authentication provider. Obviously ultimately the OAuth2 work will provide a better solution here, but we'll need to do something in the meantime.

Can we just refuse to issue a refresh token (falling back to the existing behaviour) if a `session_lifetime` is specified?